### ISSUE TITLE
feat: algorithms for cluster selection

### ISSUE THREAD

**Issue by vsoch** [converged-computing/rainbow#15]
To make it easy to "switch in" and out algorithms and run experiments, we should be able to customize this via the config, with different algorithms defined as plugins (akin to the graph database backends) and their specific parameters customizable in the config. As an example, here is what the default might look like:

```yaml
scheduler:
    secret: chocolate-cookies
    name: rainbow-cluster
    algorithm:
     name: random
     options:
        key: value

graphdatabase:
    name: memory
    options:
      host: "127.0.0.1:50051"

clusters:
  - name: keebler
    token: rainbow
```

The above also would address a current design flaw of expecting to ask the graph database for a list of contender clusters, and then rely on rainbow to ask _again_ and introduce a problem of concurrency / agreement between the clusters, because rainbow would have enough information to decide upon receiving the list (this strategy will need to be further discussed, but largely there should be summary metrics of state that rainbow can keep about the clusters to supplement the graph).

The above organization can be tweaked, but the high level idea is that it is saying "rainbow should provide the summary metrics to algorithm `random` that can accept the input parameters `key=value` from the user, and use this to determine the assignment. This means that the flow of information goes as follows:

```
client -> submit job --> graph database --> returns to client --> send to rainbow --> rainbow decides --> assigns to cluster
```
instead of:

```
client -> submit job --> graph database --> returns to client --> send to rainbow --> rainbow asks each cluster --> waits for decision/agreement --> rainbow assigns based on agreement --> assigns to cluster
```
The first model has less potential for jobs getting lost, and less error with respect to needing for all clusters to respond before making a decision. The "random" algorithm can be default, and would not require any parameters, but merely just randomly select from the list. Likely this will be a good base case for experiments where we are trying to prove that some selection algorithm is better than chosing at random. This design will also allow me to move forward with subsystem thinking (which is a bit of a catch 22 because likely we need that for the algorithms in the first place)!

**Comment by vsoch** [comment]
This will be closed with #14

**Comment by vsoch** [comment]
This was done, forgot to close it.
